subsys/shell/shell_uart.c

```diff
@@ -7,37 +7,125 @@
 #include <shell/shell_uart.h>
 #include <uart.h>
 #include <init.h>
+#include <logging/log.h>
 
-SHELL_UART_DEFINE(shell_transport_uart);
+#define LOG_MODULE_NAME shell_uart
+LOG_MODULE_REGISTER(shell_uart);
+
+SHELL_UART_DEFINE(shell_transport_uart,
+		  CONFIG_SHELL_BACKEND_SERIAL_TX_RING_BUFFER_SIZE,
+		  CONFIG_SHELL_BACKEND_SERIAL_RX_RING_BUFFER_SIZE);
 SHELL_DEFINE(uart_shell, "uart:~$ ", &shell_transport_uart, 10,
 	     SHELL_FLAG_OLF_CRLF);
 
-static void timer_handler(struct k_timer *timer)
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static void uart_rx_handle(const struct shell_uart *sh_uart)
 {
-	struct shell_uart *sh_uart =
-			CONTAINER_OF(timer, struct shell_uart, timer);
+	u8_t *data;
+	u32_t len;
+	u32_t rd_len;
+	bool new_data = false;
+
+	do {
+		len = ring_buf_put_claim(sh_uart->rx_ringbuf, &data,
+					 sh_uart->rx_ringbuf->size);
+		rd_len = uart_fifo_read(sh_uart->ctrl_blk->dev, data, len);
+
+		if (rd_len) {
+			new_data = true;
+		}
+
+		ring_buf_put_finish(sh_uart->rx_ringbuf, rd_len);
+	} while (rd_len && (rd_len == len));
+
+	if (new_data) {
+		sh_uart->ctrl_blk->handler(SHELL_TRANSPORT_EVT_RX_RDY,
+					   sh_uart->ctrl_blk->context);
+	}
+}
 
-	if (uart_poll_in(sh_uart->dev, sh_uart->rx) == 0) {
-		sh_uart->rx_cnt = 1;
-		sh_uart->handler(SHELL_TRANSPORT_EVT_RX_RDY, sh_uart->context);
+static void uart_tx_handle(const struct shell_uart *sh_uart)
+{
+	u32_t len;
+	u8_t *data;
+	int err;
+	struct device *dev = sh_uart->ctrl_blk->dev;
+
+	len = ring_buf_get_claim(sh_uart->tx_ringbuf, &data,
+				 sh_uart->tx_ringbuf->size);
+	if (len) {
+		len = uart_fifo_fill(dev, data, len);
+		err = ring_buf_get_finish(sh_uart->tx_ringbuf, len);
+		assert(err == 0);
```

Comment:
1. We shall use `__ASSERT_NO_MSG`
2. Don't you get any compiler warning about unused variable `err` when assert is off?